features/step_definitions/section_explicit_waiting_steps.rb

```diff
@@ -4,34 +4,57 @@
   @test_site.slow.first_section(wait: upper_bound_delay)
 end
 
-When('I wait for the section that takes a while to vanish') do
-  @test_site.vanishing.wait_until_delayed_section_invisible
+When('I wait for the section that takes a while to vanish when calling from {is_parent_section} section') do |from_parent_section|
+  @test_site.vanishing.then do |vanishing|
+    if from_parent_section
+      vanishing.wait_until_delayed_section_invisible
+    else
+      vanishing.delayed_section.wait_until_invisible
+    end
+  end
 end
 
 Then("an exception is raised when I wait for a section that won't appear") do
   expect { @test_site.slow.first_section(wait: lower_bound_delay) }
     .to raise_error(Capybara::ElementNotFound)
 end
 
-Then('an error is raised when waiting for the section to vanish') do
-  expect { @test_site.vanishing.wait_until_container_invisible }.to raise_error(SitePrism::ElementInvisibilityTimeoutError)
+Then('an error is raised when waiting for the section to vanish when calling from {is_parent_section} section') do |from_parent_section|
+  expect do
+    @test_site.vanishing.then do |vanishing|
+      from_parent_section ? vanishing.wait_until_container_invisible : vanishing.container.wait_until_invisible
+    end
+  end.to raise_error(SitePrism::ElementInvisibilityTimeoutError)
 end
 
 Then('the section is no longer visible') do
   expect(@test_site.vanishing.delayed_section).not_to be_visible
 end
 
-When('I wait an overridden time for the section to vanish') do
-  @test_site.vanishing.wait_until_delayed_section_invisible(wait: upper_bound_delay)
+When('I wait an overridden time for the section to vanish when calling from {is_parent_section} section') do |from_parent_section|
+  @test_site.vanishing.then do |vanishing|
+    if from_parent_section
+      vanishing.wait_until_delayed_section_invisible(wait: upper_bound_delay)
+    else
+      vanishing.delayed_section.wait_until_invisible(wait: upper_bound_delay)
+    end
+  end
 end
 
 Then('the slow section appears') do
   expect(@test_site.slow).to have_first_section
 end
 
-Then('an error is raised when waiting an overridden time for the section to vanish') do
-  expect { @test_site.vanishing.wait_until_container_invisible(wait: time_delay) }
-    .to raise_error(SitePrism::ElementInvisibilityTimeoutError)
+Then('an error is raised when waiting an overridden time for the section to vanish when calling from {is_parent_section} section') do |from_parent_section|
+  expect do
```

Comment:
Instead of wrapping the entire assertion inside a conditional block, just render your conditional block with 2 diff assertions, it reads a lot nicer